### PR TITLE
perf: Add the migration part to local storage performance improvement

### PR DESCRIPTION
**Pull request by yuhao900914**
### Checklist

* [x] Does your PR title have the correct [title format](https://github.com/amplitude/Amplitude-iOS/blob/main/CONTRIBUTING.md#pr-commit-title-conventions)?
* Does your PR have a breaking change?:

### CODE REVIEW

File: Sources/Amplitude/Amplitude.m [hunk 1]
```diff
@@ -255,21 +258,6 @@ - (instancetype)initWithInstanceName:(NSString *)instanceName {
                 AMPLITUDE_LOG(@"Loaded from %@", _propertyListPath);
             }
 
-            // update database if necessary
-            int oldDBVersion = 1;
-            NSNumber *oldDBVersionSaved = [self->_propertyList objectForKey:DATABASE_VERSION];
```
Comment (qingzhuozhen) [inline Sources/Amplitude/Amplitude.m:260]: Is this old migration? Just curious is there any plan about the `DATABASE_VERSION`?

File: Sources/Amplitude/Amplitude.m [hunk 2]
```diff
@@ -284,6 +272,49 @@ - (instancetype)initWithInstanceName:(NSString *)instanceName {
...
+- (void)migrateToFileStorage {
+     NSString *databaseDirectory = [AMPUtils platformDataDirectory];
+     NSString *databasePath = [databaseDirectory stringByAppendingPathComponent:@"com.amplitude.database"];
+    if (![self->_instanceName isEqualToString:kAMPDefaultInstance]) {
+        databasePath = [NSString stringWithFormat:@"%@_%@", databasePath, self.instanceName];
+    }
+    NSString *fileStoragePath =  [AMPStorage getAppStorageAmpDir:self.instanceName];
+
+    BOOL firstMigration = ![[NSFileManager defaultManager] fileExistsAtPath:fileStoragePath isDirectory: true];
+    BOOL needDBMigrate =  ([[NSFileManager defaultManager] fileExistsAtPath:databasePath] && firstMigration) ? true : false;
```
Comment (qingzhuozhen) [inline Sources/Amplitude/Amplitude.m:284]: So clarify the logic a bit, firstMigration is true when there is no new file storage, and that's also why getAppStorageAmpDir is changed to public. And needDBMigrate is checking if we have the database now?

Comment (yuhao900914) [reply Sources/Amplitude/Amplitude.m:284]: Yes, you are right.

File: Sources/Amplitude/Amplitude.m [hunk 3]
```diff
@@ -284,6 +272,49 @@ - (instancetype)initWithInstanceName:(NSString *)instanceName {
...
+    if (!needDBMigrate) return;
+    int eventCount = [self.dbHelper getEventCount];
+    if (eventCount > 0) {
+        NSArray *events = [self.dbHelper getEvents: -1 limit: -1];
+        NSDictionary *eventsDict = [self mergeEventsAndIdentifys:events identifys:nil numEvents:eventCount];
+        long long maxEventId = [[eventsDict objectForKey:MAX_EVENT_ID] longLongValue];
+        for(NSString *event in events) {
+            [AMPStorage storeEvent:event instanceName:self.instanceName];
```
Comment (qingzhuozhen) [inline Sources/Amplitude/Amplitude.m:292]: So when would we actually load this event? After we save it to the file, in logEvent it would not read from the file right? Per @dantetam's last pr, it will only load the file when it gets to foreground? Would we just happen to truncate or remove the file but not sending the old events?

Comment (yuhao900914) [reply Sources/Amplitude/Amplitude.m:292]: I think that should not happen. 
1. in the storeEvent, we have logic that if we already create the txt file we only append events at the end of the file. 
2. we only remove the files after we load the event to the server and get success response.

File: Sources/Amplitude/Amplitude.m [hunk 4]
```diff
@@ -284,6 +272,49 @@ - (instancetype)initWithInstanceName:(NSString *)instanceName {
...
+        NSDictionary *identifysDict = [self mergeEventsAndIdentifys:nil identifys:identifys numEvents:identifyCount];
+        long long maxIdentifyId = [[identifysDict objectForKey:MAX_IDENTIFY_ID] longLongValue];
+        for(NSString *identify in identifys) {
+            [AMPStorage storeIdentify:identify instanceName:self.instanceName];
+        }
+        [self.dbHelper removeIdentifys:maxIdentifyId];
+    }
+
+    // migrate for store database table
+    [self->_defaultDataStorage setObject:[self.dbHelper getValue:USER_ID] forKey:USER_ID];
```
Comment (haoliu-amp) [inline Sources/Amplitude/Amplitude.m:309]: This is very dangerous since our user is likely to use NSUserDefaults too. They might have the same key for USER_ID and other keys below.

Comment (haoliu-amp) [reply Sources/Amplitude/Amplitude.m:309]: Need to add prefix for the key. 
e.g. amplitude.user_id OR amplitude_user_id

**Review by haoliu-amp — CHANGES_REQUESTED** [review]

File: Sources/Amplitude/Amplitude.m (hunk 4 above)
Comment (haoliu-amp) [reply Sources/Amplitude/Amplitude.m:309]: better update the constant and create constants for OLD_USER_ID

Comment (haoliu-amp) [reply Sources/Amplitude/Amplitude.m:309]: Also do we need to consider instanceName here?

File: Sources/Amplitude/Amplitude.m [hunk 5]
```diff
@@ -284,6 +272,49 @@ - (instancetype)initWithInstanceName:(NSString *)instanceName {
...
 
+- (void)migrateToFileStorage {
+     NSString *databaseDirectory = [AMPUtils platformDataDirectory];
+     NSString *databasePath = [databaseDirectory stringByAppendingPathComponent:@"com.amplitude.database"];
+    if (![self->_instanceName isEqualToString:kAMPDefaultInstance]) {
+        databasePath = [NSString stringWithFormat:@"%@_%@", databasePath, self.instanceName];
+    }
+    NSString *fileStoragePath =  [AMPStorage getAppStorageAmpDir:self.instanceName];
+
+    BOOL firstMigration = ![[NSFileManager defaultManager] fileExistsAtPath:fileStoragePath isDirectory: true];
```
Comment (haoliu-amp) [inline Sources/Amplitude/Amplitude.m:283]: Why not checking this at the very beginning without executing other logics?

File: Sources/Amplitude/AMPStorage.h [hunk 6]
```diff
@@ -1,5 +1,5 @@
 //
-//  Amplitude.m
+//  Amplitude.h
```
Comment (haoliu-amp) [inline Sources/Amplitude/AMPStorage.h:2]: AMPStorage.h

Comment (haoliu-amp) [reply Sources/Amplitude/AMPStorage.h:2]: Also please correct it for the `m` file too. AMPStorage.m

File: Sources/Amplitude/Amplitude.m [hunk 7]
```diff
@@ -1519,7 +1613,7 @@ - (NSDate *)currentTime {
 }
 
 - (void)printEventsCount {
```
Comment (haoliu-amp) [inline Sources/Amplitude/Amplitude.m:1615]: Do we need this function?

File: Sources/Amplitude/Amplitude.m [hunk 8]
```diff
@@ -114,6 +116,12 @@ @interface Amplitude ()
 static NSString *const USER_ID = @"user_id";
 static NSString *const SEQUENCE_NUMBER = @"sequence_number";
 
+static NSString *const OLD_USER_ID = @"user_id";
```
Comment (haoliu-amp) [inline Sources/Amplitude/Amplitude.m:119]: OLD and NEW have the same key?

**Review by haoliu-amp — CHANGES_REQUESTED** [review]

File: Sources/Amplitude/Amplitude.m (hunk 8 above)
Comment (qingzhuozhen) [reply Sources/Amplitude/Amplitude.m:119]: I believe in new storage @yuhao900914 is trying to use `getDataStorageKey` for new key, but I guess we don't need to define the old const for it.

File: Sources/Amplitude/Amplitude.m [hunk 9]
```diff
@@ -284,6 +280,82 @@ - (instancetype)initWithInstanceName:(NSString *)instanceName {
     return self;
 }
 
+- (NSString *)getDataStorageKey:(NSString *)key {
+    NSString *dataStorageKey = [NSString stringWithFormat:@"%s_%@_%@_%@", "amplitude", _bundleIdentifier ,self.instanceName, key];
```
Comment (qingzhuozhen) [inline Sources/Amplitude/Amplitude.m:284]: nit: the space in `_bundleIdentifier ,self.instanceName`

File: Sources/Amplitude/Amplitude.m [hunk 10]
```diff
@@ -284,6 +280,82 @@ - (instancetype)initWithInstanceName:(NSString *)instanceName {
...
+    return dataStorageKey;
+}
+
+- (void)migrateToFileStorage {
+    NSString *fileStoragePath =  [AMPStorage getAppStorageAmpDir:self.instanceName];
+    BOOL isDir;
+    [[NSFileManager defaultManager] fileExistsAtPath:fileStoragePath isDirectory:&isDir];
+    if (isDir) return;
+
+    NSString *databaseDirectory = [AMPUtils platformDataDirectory];
```
Comment (qingzhuozhen) [inline Sources/Amplitude/Amplitude.m:294]: nit: in theory these two check could be moved to storage and dbhelper as a check method to make it more clear. But since it is only for migration, I guess it is fine to leave it for now.

File: Sources/Amplitude/Amplitude.m (hunk 9 above)
Comment (haoliu-amp) [reply Sources/Amplitude/Amplitude.m:284]: Make it to be a static func

File: Sources/Amplitude/Amplitude.m (hunk 8 above)
Comment (haoliu-amp) [reply Sources/Amplitude/Amplitude.m:119]: I see, makes sense.

File: Sources/Amplitude/Amplitude.m (hunk 9 above)
Comment (haoliu-amp) [reply Sources/Amplitude/Amplitude.m:284]: My main concern for having this function is that other folks might not know to call this at all. However, I have no better solution atm.

File: Sources/Amplitude/Amplitude.m (hunk 7 above)
Comment (yuhao900914) [reply Sources/Amplitude/Amplitude.m:1615]: In the header file, It says this method is used for debugging method to find out how many events are being stored locally on the device. So I think that's still valid?

File: Sources/Amplitude/Amplitude.m (hunk 8 above)
Comment (yuhao900914) [reply Sources/Amplitude/Amplitude.m:119]: Yes, the key for new storage is using `getDataStorageKey`, change to OLD_XXX is for making it more clear those keys are points to the old key for database.

File: Sources/Amplitude/AMPStorage.m [hunk 11]
```diff
@@ -112,7 +112,14 @@ + (NSMutableArray *)getEventsFromDisk:(NSString *)path {
 }
 
 + (NSDictionary *)JSONFromFile:(NSString *)path {
-    NSData *data = [NSData dataWithContentsOfFile:path];
+    NSData *data;
```
[the remark this answers is not in the document]
Comment (dantetam) [reply Sources/Amplitude/AMPStorage.m:115]: That's a good catch. Good use of the Objective C syntax too.

File: Sources/Amplitude/Amplitude.m (hunk 1 above)
Comment (dantetam) [reply Sources/Amplitude/Amplitude.m:260]: I thought we were completely removing the old migration. I also removed a very old (5 years old) migration from the Android DB.

File: Sources/Amplitude/Amplitude.m [hunk 12]
```diff
@@ -284,6 +280,82 @@ - (instancetype)initWithInstanceName:(NSString *)instanceName {
...
+    NSString *databasePath = [databaseDirectory stringByAppendingPathComponent:@"com.amplitude.database"];
```
Comment (dantetam) [inline Sources/Amplitude/Amplitude.m:295]: We should standardize the string here, "com.amplitude.database". I believe other people like Segment put them in Constants file. This is just for cleaner code when someone updates it later.

File: Sources/Amplitude/Amplitude.m [hunk 13]
```diff
@@ -284,6 +280,82 @@ - (instancetype)initWithInstanceName:(NSString *)instanceName {
...
+    if (![self->_instanceName isEqualToString:kAMPDefaultInstance]) {
+        databasePath = [NSString stringWithFormat:@"%@_%@", databasePath, self.instanceName];
+    }
+    BOOL hasDatabase =  [[NSFileManager defaultManager] fileExistsAtPath:databasePath];
+    if (!hasDatabase) return;
+
+    int eventCount = [self.dbHelper getEventCount];
+    if (eventCount > 0) {
+        NSMutableArray *events = [self.dbHelper getEvents: -1 limit: -1];
```
Comment (dantetam) [inline Sources/Amplitude/Amplitude.m:304]: nit: no space between parameter names and values in objective C

File: Sources/Amplitude/Amplitude.m [hunk 14]
```diff
@@ -284,6 +280,82 @@ - (instancetype)initWithInstanceName:(NSString *)instanceName {
...
+            NSError *error = nil;
+            NSData *jsonData = [NSJSONSerialization dataWithJSONObject:[AMPUtils makeJSONSerializable:event] options:0 error:&error];
+            if (error != nil) {
+                AMPLITUDE_ERROR(@"ERROR: could not JSONSerialize event type: %@", error);
+                continue;
+            }
+            NSString *jsonString = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
+            [AMPStorage storeEvent:jsonString instanceName:self.instanceName];
+        }
+       NSDictionary *eventsDict = [self mergeEventsAndIdentifys:events identifys:nil numEvents:eventCount];
```
Comment (dantetam) [inline Sources/Amplitude/Amplitude.m:316]: Why are we merging here if we're doing the identify events separately on the next few lines (321)?

Comment (yuhao900914) [reply Sources/Amplitude/Amplitude.m:316]: Good catch. You are right, we only need a maxEventId here for removing all events in the database.  Will fix this.

**Comment by qingzhuozhen** [comment]
I see some error in tests about xcodebuild, can we fix it? @yuhao900914 would these errors affecting further build or runs?

File: Sources/Amplitude/Amplitude.m [hunk 15]
```diff
@@ -284,6 +271,71 @@ - (instancetype)initWithInstanceName:(NSString *)instanceName {
...
+                continue;
+            }
+            [AMPStorage storeIdentify:jsonString instanceName:self.instanceName];
+        }
+        long long maxIdentifyId = [[[identifys lastObject] objectForKey:@"event_id"] longLongValue];
+        [self.dbHelper removeIdentifys:maxIdentifyId];
+    }
+
+    if (eventCount > 0 || identifyCount > 0) {
+        [self uploadEvents];
```
Comment (qingzhuozhen) [inline Sources/Amplitude/Amplitude.m:326]: Just notice we are calling `uploadEvents` now, would it send the events actually? Or are we writing the same events to the file again?

File: Sources/Amplitude/Amplitude.m [hunk 16]
```diff
@@ -703,24 +755,23 @@ - (void)uploadEventsWithLimit:(int)limit {
...
         if ([self optOut] || self->_offline) {
             self->_updatingCurrently = NO;
             [self endBackgroundTaskIfNeeded];
             return;
         }
 
-        int eventCount = [self->_eventsBuffer count];
-        int numEvents = limit > 0 ? fminl(eventCount, limit) : eventCount;
+        long eventCount = [self->_eventsBuffer count];
+        long numEvents = limit > 0 ? fminl(eventCount, limit) : eventCount;
```
Comment (qingzhuozhen) [inline Sources/Amplitude/Amplitude.m:766]: good job to cover the lossy conversions.

File: Amplitude.xcodeproj/project.pbxproj [hunk 17]
```diff
@@ -1264,6 +1286,10 @@
 				DYLIB_COMPATIBILITY_VERSION = 1;
 				DYLIB_CURRENT_VERSION = 1;
 				DYLIB_INSTALL_NAME_BASE = "@rpath";
+				FRAMEWORK_SEARCH_PATHS = (
+					"$(inherited)",
```
Comment (qingzhuozhen) [inline Amplitude.xcodeproj/project.pbxproj:1290]: Is this all generated from xcode?

Comment (dantetam) [reply Amplitude.xcodeproj/project.pbxproj:1290]: Looks like it. xcode settings too. Might want to take a look at `FRAMEWORK_SEARCH_PATHS`, is this something you saw before @yuhao900914 ?

**Review by qingzhuozhen — APPROVED** [review]
Let's get on the tests and make sure we cover the corner cases

File: Amplitude.xcodeproj/project.pbxproj (hunk 17 above)
Comment (yuhao900914) [reply Amplitude.xcodeproj/project.pbxproj:1290]: Yes, it's auto generated by xcode. When doing the development, we need to open the Amplitude.xcworkspace instead of the whole folder. Then you will find, the AMPStorage files are gone. It cannot build the file when runing the test.
